### PR TITLE
Add support for `weights_only` flag when loading state_dict

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
This is to enable loading a state_dict with wrapper tensor subclasses (used in torchao to for quantized weights)

Test Plan:
tested locally with torchao weights, also need https://github.com/huggingface/transformers/pull/32306:
```
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from transformers import TorchAoConfig
from torchao.utils import benchmark_model
import torchao

DEVICE_TYPE = "cuda"

def init_model_and_benchmark(model_id, torch_dtype=torch.bfloat16, quantization_config=None):
    tokenizer = AutoTokenizer.from_pretrained(model_id)
    if quantization_config is not None:
        model = AutoModelForCausalLM.from_pretrained(model_id, device_map=DEVICE_TYPE, torch_dtype=torch.bfloat16, quantization_config=quantization_config)
    else:
        model = AutoModelForCausalLM.from_pretrained(model_id, device_map=DEVICE_TYPE, torch_dtype=torch.bfloat16, weights_only=False)

    # sanity check: run the model
    input_text = "What are we having for dinner?"
    input_ids = tokenizer(input_text, return_tensors="pt").to(DEVICE_TYPE)
    output = model.generate(**input_ids, max_new_tokens=1000)
    print(tokenizer.decode(output[0], skip_special_tokens=True))

    NUM_WARMUP = 1
    NUM_RUNS = 5

    if quantization_config is not None:
        torchao.quantization.utils.recommended_inductor_config_setter()

    model = torch.compile(model, mode="max-autotune")

    benchmark_model(model.generate, NUM_WARMUP, kwargs=input_ids, device_type=DEVICE_TYPE)
    print("running benchmark")
    results = benchmark_model(model.generate, NUM_RUNS, kwargs=input_ids, device_type=DEVICE_TYPE)
    return model, results

model_id = "jerryzh168/test-model"
torchao.quantization.utils.recommended_inductor_config_setter()
bf16_model, bf16_time = init_model_and_benchmark(model_id)
print(f"bf16: {bf16_time}")
```

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -3432,6 +3438,7 @@ def from_pretrained(
                     os.path.join(pretrained_model_name_or_path, subfolder, _add_variant(WEIGHTS_NAME, variant))
                 ):
                     # Load from a PyTorch checkpoint
+                    print("pytorch checkpoint")
```
Comment (amyeroberts) [inline src/transformers/modeling_utils.py:3441]: If we're emitting to the terminal, it should be using `logging`. I don't think this would be very informative for the user however

```suggestion
```

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this support! 

Overall, I think this looks good. Let's get a second 👍 from @ArthurZucker too as it's touching core code

**Comment by jerryzh168** [comment]
I learned from @mikaylagawarecki that `weights_only` is going to be set to True by default in the future. but this flag could still be helpful if people are using older versions of pytorch and want to use torchao I think

**Comment by jerryzh168** [comment]
Hi @ArthurZucker can you take a look at this PR?

**Comment by jerryzh168** [comment]
please let me know if we want to add a test, but load test is a bit harder to write because it needs to upload some models like: https://github.com/huggingface/transformers/blob/abbffc4525566a48a9733639797c812301218b83/tests/test_modeling_utils.py#L930

**Review by ArthurZucker — APPROVED** [review]
LGTM sorry for being late on the review here!

**Comment by ArthurZucker** [comment]
Can you rebase on main and resolve the conflicts?

**Comment by jerryzh168** [comment]
thanks for the review @ArthurZucker and @amyeroberts just updated the PR, please feel free to merge when the CI is green

**Comment by amyeroberts** [comment]
@jerryzh168 Could you run `make fix-copies` and push the changes? This should resolve the failing quality checks

**Comment by jerryzh168** [comment]
thanks, updated, please take a look again @amyeroberts

File: src/transformers/models/olmoe/modeling_olmoe.py
Comment (ArthurZucker) [inline src/transformers/models/olmoe/modeling_olmoe.py:1]: this unrelated change should not be included here ! 🤗

**Review by ArthurZucker — COMMENTED** [review]
let's revert unrelated changes and let's go

File: src/transformers/models/olmoe/modeling_olmoe.py
Comment (jerryzh168) [reply src/transformers/models/olmoe/modeling_olmoe.py:1]: @ArthurZucker this is fix from running `make fix-copies` 😅

Comment (ArthurZucker) [reply src/transformers/models/olmoe/modeling_olmoe.py:1]: ok no worries then let's merge!

**Review by ArthurZucker — APPROVED** [review]
Sorry for the delay would kindly ask to rebase to make sure we are up to date as other things were merged !

**Comment by ArthurZucker** [comment]
Thanks once more @jerryzh168
